### ISSUE TITLE
How do I format Ruby files?

### ISSUE THREAD

**Issue by khuongduybui** [svipas/vscode-prettier-plus#8]
I installed Prettier+ latest version on latest version of VSCode stable, open a .rb file and run Format Document.

Edit (again): I installed prettier and @prettier/plugin-ruby locally and still cannot issue the format document command (No formatter installed error).

**Comment by svipas** [comment]
If you can share the file with content and full name including extension for e.g. `ruby.rb`

**Comment by pjforde1978** [comment]
I have the same issue, "no formatter installed".

I am using:

```
  "[ruby]": {
    "editor.defaultFormatter": "svipas.prettier-plus"
  },
```

in my settings.js and see no result from a Format Document operation OR the format-on-save.

**Comment by svipas** [comment]
> If you can share the file with content and full name including extension for e.g. ruby.rb

**Comment by pjforde1978** [comment]
Github won't let me attach .rb files, but here's a direct link to a file I'm working on.

https://raw.githubusercontent.com/hopsoft/stimulus_reflex/master/lib/stimulus_reflex/channel.rb

Note that the issue both of us have is that VS Code suggests that there is no Ruby formatter installed.

**Comment by svipas** [comment]
First problem I encounter when I have installed `@prettier/plugin-ruby` it doesn't detect automatically Ruby files. So basically I will try to detect Prettier plugins and enable them automatically for that type of file extension for e.g. in this case Ruby (`.rb`) so you don't need to do add any: `"editor.defaultFormatter": "svipas.prettier-plus"`. Second problem is it doesn't format code at all with these plugins. I would take a look and I will try to fix these issues. Thank you for bug report!

**Comment by pjforde1978** [comment]
Well, the main reason that I have a [ruby] block in my settings is that I want to use a completely different formatter for my javascript code. As you said in the README, you can run into conflicts when you have multiple formatters trying to take control. Does this change your opinion? I want different formatters for different languages happily co-existing.

I admit that I'm not familiar with the internals of VSCode so I don't know if I understand what you mean about the @prettier/plugin-ruby library. Is that something extra that I have to install to make this solution work?

**Comment by svipas** [comment]
@pjforde1978 You can use different formatters for different languages by using that setting if you want. What I meant is you don't need to additionally add that setting to format ruby or other file.

Prettier formats a lot of extensions, but to format some of them for e.g. like Ruby you should have `@prettier/plugin-ruby`. You can check whole list in https://prettier.io (Community Plugins).

I'm developing solution right now, I will also update documentation to explain how to format code with Ruby.

**Comment by pjforde1978** [comment]
I really appreciate you following-up on this.

Documentation is key. The honest truth is that while you are a domain expert, most of the people using your plugin will have no thought to wonder if they are missing a vital step if there is no documentation to suggest more is required.

Anyhow, are you proposing that I install the prettier gem, install the npm package (globally?) or something else?

You could add a language-specific tips section to your README:

`yarn global add prettier @prettier/plugin-ruby`

**Comment by khuongduybui** [comment]
Where are we with this? Will the fix be included in your upcoming big release? Pretty (prettier) please?

**Comment by svipas** [comment]
@khuongduybui Yes, of course, for now I have other work to do, but I will release a new version this week with plugins support and ESLint 6.

**Comment by svipas** [comment]
Good news guys, I added support for official plugins and it even formats untitled files if you have installed plugin locally in your project or in any workspace project. You don't need to use any config in VS Code because now it automatically registers these languages. I refactored all `README.md` which now includes proper language support list, better documentation for integrations, plugins, etc.

**Comment by khuongduybui** [comment]
Awesome I'll check it out right away

**Comment by khuongduybui** [comment]
oh but you haven't released the update version have you?

**Comment by svipas** [comment]
Just wait a little bit, I'm currently writing tests and preparing for a new 3.0.0 release (current version is 2.0.0). I will close this issue when it's done, it will take about 30 minutes from now.

**Comment by svipas** [comment]
New version is published. Check documentation of plugins here: https://github.com/svipas/vscode-prettier-plus#plugins

/cc @khuongduybui @pjforde1978

**Comment by khuongduybui** [comment]
I just checked it out. Worked great even in Remote WSL! Thanks so much for fixing this. `prettier-vscode` claims to support plugins but I could never get it to work.

**Comment by pjforde1978** [comment]
I really want to try this out, but since I first posted my team has settled upon passing everything through `standardrb --fix` and I'm worried that adding prettier might occasionally create format conflicts. I look forward to trying this out on a personal project soon.

**Comment by svipas** [comment]
@khuongduybui Thanks for checking it out in Remote WSL.

It should work in these scenarios:

1. In untitled files if opened project contains installed plugin locally.
2. You can open several workspaces and install plugin only in one and it should format all workspaces and untitled files as well.
3. In remote if project contains installed plugin locally.
4. Normally with one opened folder and installed plugin locally.

Basically it should work in all cases and what you need to do is only install plugin locally.

@pjforde1978 No problem, I hope it will help in your personal projects.

**Comment by khuongduybui** [comment]
Why does it need to be installed locally though? Why can't I yarn global add and expect it to always work?

**Comment by svipas** [comment]
Because it does searching across your opened workspaces. To do it globally it's way harder because of different package managers and OSs. I'm trying to simplify setup as easier as possible, but this is best solution for now. After plugins will be stable since they are beta for now in Prettier I will try to come with better solution like bundling them or trying to implement search globally instead of workspaces.

**Comment by khuongduybui** [comment]
What if it's simply in the path and you don't have to worry where it's from?

Duy K. Bui,
+1 571 239 6002

________________________________

**Comment by svipas** [comment]
This is kinda interesting even to try to implement to resolve it automatically. You can open a new issue with this feature request, maybe I will try to resolve it in future. You could create a issue with 2 suggestions like providing path by yourself via config or try to resolve it automatically where global packages exist.
